### PR TITLE
[IMP] sale_purchase: add SO's AA to PO when subcontracting services

### PR DESCRIPTION
**Pull request by ravb-odoo**
This commit enhances the functionality of the sale and purchase order process to automate the setting of the analytical distribution in the purchase order line. Now, the analytic account linked with the Sales Order (SO) is automatically assigned to the purchase order when subcontracting services.

task-3256865

### CODE REVIEW

**Review by pkyriakou — COMMENTED** [review]
Thank you Ravi :100: 
I had a question, at first glance, it seems like your code works when we create a new analytic account but what happens when we put an already existing account in the sale order ?

**Comment by ravb-odoo** [comment]
Hello @pkyriakou 
I have made the required adjustments, and now this solution works correctly for both scenarios.
Thank You:)

File: addons/sale_purchase/models/sale_order.py [hunk 1]
```diff
@@ -17,12 +17,25 @@ def _compute_purchase_order_count(self):
...
+        purchase_order_lines = purchase_orders.order_line.filtered(lambda line: line.sale_order_id == self)
+        for order_line in purchase_order_lines:
+            order_line.analytic_distribution = {self.analytic_account_id.id: 100}
+
     def _action_confirm(self):
         result = super(SaleOrder, self)._action_confirm()
         for order in self:
             order.order_line.sudo()._purchase_service_generation()
+            if order.analytic_account_id:
+                self._set_analytic_account()
```
Comment (pkyriakou) [inline addons/sale_purchase/models/sale_order.py:31]: I don't think this check is needed since the order in self will be given an ``analytic_account_id`` from the super call.

File: addons/sale_purchase/models/sale_order.py [hunk 2]
```diff
@@ -17,12 +17,25 @@ def _compute_purchase_order_count(self):
...
         return result
 
+    def _create_analytic_account(self, prefix=None):
+        super()._create_analytic_account()
+        # set analytic_account fron SO to purchase order line
+        self._set_analytic_account()
+
```
Comment (pkyriakou) [inline addons/sale_purchase/models/sale_order.py:38]: Same with this, since the set_analytic_account is called in the action confirm, I dont think it needs to be called in the ``_create_analytic_account`` method as well.

**Review by pkyriakou — COMMENTED** [review]
Thank you for your changes @ravb-odoo 
I left some comments below :100:

File: addons/sale_purchase/models/sale_order.py (hunk 2 above)
Comment (ravb-odoo) [reply addons/sale_purchase/models/sale_order.py:38]: Hello @pkyriakou 
When the sale_project module is installed, its action_confirm function is triggered first. Subsequently, the action_confirm function of the sale_purchase module is invoked through the inheritance mechanism. This sequential execution results in an issue where the analytic account, which is initially created in the sale_project module, is not set correctly. Consequently, it becomes necessary to call the set_analytic_account method twice to ensure the proper configuration of the analytic account in both manual and automatic scenarios.

Thank You:)

Comment (pkyriakou) [reply addons/sale_purchase/models/sale_order.py:38]: It still seems strange to me :thinking: but we can either way move ahead with it

**Review by pkyriakou — COMMENTED** [review]
Thank you very much @ravb-odoo :100: 
I have nothing else to comment, LGTM :green_circle: :partying_face:

**Review by xavierbol — COMMENTED** [review]
Hello @ravb-odoo 

Thanks for your work! 🙂 

May I ask you to rebase your branch to be up to date and also add a unit test? 

Thanks in advance! 🙂

**Comment by ravb-odoo** [comment]
Hello @xavierbol ,
I've included a unit test and also rebase a branch.

Thank You :)

File: addons/sale_purchase/models/sale_order.py [hunk 3]
```diff
@@ -17,12 +17,25 @@ def _compute_purchase_order_count(self):
...
+    def _set_analytic_account(self):
+        purchase_orders = self._get_purchase_orders()
+        purchase_order_lines = purchase_orders.order_line.filtered(lambda line: line.sale_order_id == self)
+        for order_line in purchase_order_lines:
+            order_line.analytic_distribution = {self.analytic_account_id.id: 100}
+
     def _action_confirm(self):
         result = super(SaleOrder, self)._action_confirm()
         for order in self:
             order.order_line.sudo()._purchase_service_generation()
```
Comment (xavierbol) [inline addons/sale_purchase/models/sale_order.py:29]: Why not using the result of that method to get the POL? 🤔 

Or why not moving your changes inside `_purchase_service_generation` or in the methods called inside that method, for instance: `_purchase_service_prepare_line_values`?

**Review by xavierbol — COMMENTED** [review]
Thanks again for your work, I left some comments. 🙂

File: addons/sale_purchase/models/sale_order_line.py [hunk 4]
```diff
@@ -194,6 +194,7 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
             uom_id=self.product_id.uom_po_id
         )
 
+        analytic_distribution = self.analytic_distribution if self.analytic_distribution else {self.order_id.analytic_account_id.id: 100}
```
Comment (xavierbol) [inline addons/sale_purchase/models/sale_order_line.py:197]: We should also check if `self.order_id.analytic_account_id` is defined before setting the analytic_distribution with its id.

Also instead of setting False to `analytic_distribution`, we could change a little bit the method to only add `analytic_distribution` in the vals once it is needed.

Something like: 

```diff
diff --git a/addons/sale_purchase/models/sale_order_line.py b/addons/sale_purchase/models/sale_order_line.py
index 161cea9f3337..57d88867c2a8 100644
--- a/addons/sale_purchase/models/sale_order_line.py
+++ b/addons/sale_purchase/models/sale_order_line.py
@@ -201,7 +201,7 @@ class SaleOrderLine(models.Model):
         if line_description:
             name += line_description

-        return {
+        purchase_line_vals = {
             'name': name,
             'product_qty': purchase_qty_uom,
             'product_id': self.product_id.id,
@@ -212,6 +212,11 @@ class SaleOrderLine(models.Model):
             'order_id': purchase_order.id,
             'sale_line_id': self.id,
         }
+        if self.analytic_distribution:
+            purchase_line_vals['analytic_distribution'] = self.analytic_distribution
+        elif self.order_id.account_analytic_id:
+            purchase_line_vals['analytic_distribution'] = {self.order_id.account_analytic_id.id: 100.0}
+        return purchase_line_vals

     def _purchase_service_match_supplier(self, warning=True):
         # determine vendor of the order (take the first matching company and product)
```

ps: I did not test that code

**Review by xavierbol — COMMENTED** [review]
Thanks again for your work! 🙂 

I left one comment and after that it will look good to me. 🙂

File: addons/sale_purchase/models/sale_order_line.py (hunk 4 above)
Comment (ravb-odoo) [reply addons/sale_purchase/models/sale_order_line.py]: Hello @xavierbol 
I have done the changes.

Thank You :slightly_smiling_face:

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂

**Comment by xavierbol** [comment]
I will let @odoo/rd-logistics do the final review since the changes are made in their module. 🙂

**Comment by xavierbol** [comment]
Hello @odoo/rd-logistics 

I don't know if you would like to review that PR? We are ready to merge it. 🙂 

Thanks in advance! 🙂 

Have a nice day! 🙂

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 5]
```diff
@@ -305,3 +315,30 @@ def test_pol_custom_attribute(self):
         sale_order.action_confirm()
         pol = sale_order._get_purchase_orders().order_line
         self.assertEqual(pol.name, f"{self.service_purchase_1.display_name}\n\n{product_attribute.name}: {product_attribute_value.name}: {custom_value}")
+
+    def test_pol_analytic_distribution(self):
+        """Confirming SO, analytic account from SO should be set as Analytic Distribution in POL."""
+        self.sale_order_1.analytic_account_id, self.sale_order_2.analytic_account_id = self.test_analytic_account_1, self.test_analytic_account_1
```
Comment (ticodoo) [inline addons/sale_purchase/tests/test_sale_purchase.py:321]: Is one of these assignments supposed to use the `self.test_analytic_account_2`? If not, then why did you add it to the setup?

Also it's my personal preference that you do this variable assignment on 2 separate lines so it's easier to read (and makes it more obvious that you use the same account for both)

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 6]
```diff
@@ -305,3 +314,20 @@ def test_pol_custom_attribute(self):
...
+    def test_pol_analytic_distribution(self):
+        """Confirming SO, analytic account from SO should be set as Analytic Distribution in POL."""
+        self.sale_order_1.analytic_account_id = self.test_analytic_account_1
+        self.sale_order_2.analytic_account_id = self.test_analytic_account_2
+
+        self.sale_order_1.action_confirm()
+        self.sale_order_2.action_confirm()
+
+        purchase_lines_so1 = self.env['purchase.order.line'].search([('sale_line_id', 'in', self.sale_order_1.order_line.ids)])
+        purchase_line1 = purchase_lines_so1[0]
```
[the remark this answers is not in the document]
Comment (ticodoo) [reply addons/sale_purchase/tests/test_sale_purchase.py:328]: Agreed, it's better that a check fails with a reason why it failed vs a IndexError due to `purchase_lines_so1` not finding any purchase order lines. You should also check the length too before assigning this variable. Change it.

Same comment about `purchase_line2` below

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 7]
```diff
@@ -305,3 +315,30 @@ def test_pol_custom_attribute(self):
         sale_order.action_confirm()
         pol = sale_order._get_purchase_orders().order_line
         self.assertEqual(pol.name, f"{self.service_purchase_1.display_name}\n\n{product_attribute.name}: {product_attribute_value.name}: {custom_value}")
+
+    def test_pol_analytic_distribution(self):
+        """Confirming SO, analytic account from SO should be set as Analytic Distribution in POL."""
+        self.sale_order_1.analytic_account_id, self.sale_order_2.analytic_account_id = self.test_analytic_account_1, self.test_analytic_account_1
+        self.sale_order_2.order_line.analytic_distribution = {str(self.sale_order_2.analytic_account_id.id): 100}
```
Comment (ticodoo) [inline addons/sale_purchase/tests/test_sale_purchase.py:322]: Shouldn't you have 1 Sale Order with no `analytic_account_id` assigned to make sure that you're actually testing that the `sol.analytic_distribution` is correctly assigned? Or at least make this value different from the `sale_order.analytic_account_id`?

File: addons/sale_purchase/tests/test_sale_purchase.py (hunk 5 above)
Comment (ravb-odoo) [reply addons/sale_purchase/tests/test_sale_purchase.py]: Hello @ticodoo 
I modified the analytic distribution for SO2, ensuring that both the analytic distribution and analytic_account_id are distinct for SO2. Additionally, other changes have been implemented.

Thank you :slightly_smiling_face:

**Review by ticodoo — APPROVED** [review]
Good for me, thanks!

**Comment by xavierbol** [comment]
Thanks @ticodoo for your review. 😁 

Thanks again @ravb-odoo for your work! 💪🏼 

robodoo r+
